tests/test_paperqa.py

```diff
@@ -409,101 +409,137 @@ def test_extract_score():
-            citation="WikiMedia Foundation, 2023, Accessed now",
-            dockey="test",
-        )
-        answer = await docs.aget_evidence(
-            Answer(question="What is the national flag of Canada?")
-        )
-        await docs.aquery("What is the national flag of Canada?", answer=answer)
+""",
+    ],
+)
+def test_llm_read_json(example: str):
+    assert llm_read_json(example) == {"example": "json"}
+
+
+@pytest.mark.asyncio()
+async def test_chain_completion():
+    client = AsyncOpenAI()
+    llm = OpenAILLMModel(config={"model": "babbage-002", "temperature": 0.2})
+    call = llm.make_chain(
+        client,
+        "The {animal} says",
+        skip_system=True,
+    )
+    outputs = []
+
+    def accum(x):
+        outputs.append(x)
+
+    completion = await call({"animal": "duck"}, callbacks=[accum])  # type: ignore[call-arg]
+    assert completion.seconds_to_first_token > 0
+    assert completion.prompt_count > 0
+    assert completion.completion_count > 0
+    assert str(completion) == "".join(outputs)
+
+    completion = await call({"animal": "duck"})  # type: ignore[call-arg]
+    assert completion.seconds_to_first_token == 0
+    assert completion.seconds_to_last_token > 0
+
+
+@pytest.mark.asyncio()
+async def test_chain_chat():
+    client = AsyncOpenAI()
+    llm = OpenAILLMModel(
+        config={"temperature": 0, "model": "gpt-3.5-turbo", "max_tokens": 56}
+    )
+    call = llm.make_chain(
+        client,
+        "The {animal} says",
+        skip_system=True,
+    )
+    outputs = []
+
+    def accum(x):
+        outputs.append(x)
+
+    completion = await call({"animal": "duck"}, callbacks=[accum])  # type: ignore[call-arg]
+    assert completion.seconds_to_first_token > 0
+    assert completion.prompt_count > 0
+    assert completion.completion_count > 0
+    assert str(completion) == "".join(outputs)
+
+    completion = await call({"animal": "duck"})  # type: ignore[call-arg]
+    assert completion.seconds_to_first_token == 0
+    assert completion.seconds_to_last_token > 0
+
+    # check with mixed callbacks
+    async def ac(x):  # noqa: ARG001
+        pass
+
+    completion = await call({"animal": "duck"}, callbacks=[accum, ac])  # type: ignore[call-arg]
+
+
+@pytest.mark.asyncio()
+async def test_anthropic_chain():
+    try:
+        from anthropic import AsyncAnthropic
+    except ImportError:
+        return
+
+    client = AsyncAnthropic()
+    llm = AnthropicLLMModel()
+    call = llm.make_chain(
+        client,
+        "The {animal} says",
+        skip_system=True,
+    )
+
+    def accum(x):
+        outputs.append(x)
+
+    outputs: list[str] = []
+    completion = await call({"animal": "duck"}, callbacks=[accum])  # type: ignore[call-arg]
+    assert completion.seconds_to_first_token > 0
+    assert completion.prompt_count > 0
+    assert completion.completion_count > 0
+    assert str(completion) == "".join(outputs)
+    assert type(completion.text) is str  # noqa: E721
```

Comment:
```suggestion
    assert isinstance(completion.text, str)
```

And a few lines below